File: packages/core/src/rules/parameters-valid.ts
import type { RequirementWithSource, ValidationError } from "../types.js";
import type { ValidationRule } from "./types.js";

/** GRD-SYS-005: Top-level requirement fields; parameter names must not overlap these. */
const TOP_LEVEL_KEYS = new Set([
  "id",
  "title",
  "description",
  "attributes",
  "links",
  "parameters",
]);

/**
 * GRD-SYS-005: Validate that parameter names are unique within the requirement
 * and do not overlap with any other fields in the requirement.
 */
export const parametersValid: ValidationRule = (
  requirement: RequirementWithSource
): ValidationError[] => {
  const errors: ValidationError[] = [];
  const params = requirement.parameters;
  if (params == null || typeof params !== "object") return errors;

  const names = Object.keys(params);
  const seen = new Set<string>();
  for (const name of names) {
    if (seen.has(name)) {
      errors.push({
        path: requirement.sourcePath,
        message: `GRD-SYS-005: Duplicate parameter name: ${name}`,
      });
    }
    seen.add(name);
    if (TOP_LEVEL_KEYS.has(name)) {
      errors.push({
        path: requirement.sourcePath,
        message: `GRD-SYS-005: Parameter name must not overlap with requirement field: ${name}`,
      });
    }
  }
  return errors;
};

File: crates/shallgraph-core/src/rules/parameters_valid.rs
//! GRD-SYS-005: Validate that parameter names are unique within the requirement
//! and do not overlap with any other fields in the requirement.

use crate::types::{RequirementWithSource, ValidationError};

const TOP_LEVEL_KEYS: &[&str] = &[
    "id",
    "title",
    "require",
    "refinement",
    "attributes",
    "links",
    "satisfied_by",
    "verified_by",
    "parameters",
];

pub fn parameters_valid(requirement: &RequirementWithSource) -> Vec<ValidationError> {
    let mut errors = Vec::new();
    let Some(params) = &requirement.parameters else {
        return errors;
    };

    for name in params.keys() {
        if TOP_LEVEL_KEYS.contains(&name.as_str()) {
            errors.push(ValidationError::new(
                requirement.source_path.display().to_string(),
                format!(
                    "GRD-SYS-005: Parameter name must not overlap with requirement field: {name}"
                ),
            ));
        }
    }
    errors
}
